File: kotlin/test/AccountUpdaterScheduleWithResultEntityTest.kt
package voxgig.bluefinpayconexsdk.sdktest

import java.nio.file.Files
import java.nio.file.Paths

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.api.Test

import voxgig.bluefinpayconexsdk.core.Helpers
import voxgig.bluefinpayconexsdk.core.SdkEntity
import voxgig.bluefinpayconexsdk.core.BluefinPayconexSDK
import voxgig.bluefinpayconexsdk.utility.Json
import voxgig.bluefinpayconexsdk.utility.struct.Struct

@Suppress("UNCHECKED_CAST", "UNUSED_VARIABLE", "UNUSED_VALUE")
class AccountUpdaterScheduleWithResultEntityTest {

  @Test
  fun instance() {
    val testsdk = BluefinPayconexSDK.testSDK()
    val ent = testsdk.accountUpdaterScheduleWithResult(null)
    assertNotNull(ent, "expected non-null account_updater_schedule_with_result entity")
  }

  @Test
  fun basic() {
    val setup = accountUpdaterScheduleWithResultBasicSetup(null)
    // Per-op sdk-test-control.json skip.
    val mode = if (setup.live) "live" else "unit"
    for (op in arrayOf("create", "list")) {
      val reason = RunnerSupport.skipReason("entityOp", "account_updater_schedule_with_result.$op", mode)
      Assumptions.assumeTrue(
        reason == null,
        if (reason == null || "" == reason) "skipped via sdk-test-control.json" else reason,
      )
    }
    Assumptions.assumeFalse(
      setup.syntheticOnly,
      "live entity test uses synthetic IDs from fixture — set BLUEFINPAYCONEX_TEST_ACCOUNT_UPDATER_SCHEDULE_WITH_RESULT_ENTID JSON to run live",
    )
    val client = setup.client

    // CREATE
    val accountUpdaterScheduleWithResultRef01Ent = client.accountUpdaterScheduleWithResult(null)
    var accountUpdaterScheduleWithResultRef01Data: MutableMap<String, Any?> = (Helpers.toMapAny(Struct.getprop(
        Struct.getpath(setup.data, "new.account_updater_schedule_with_result"), "account_updater_schedule_with_result_ref01")) ?: linkedMapOf())
    accountUpdaterScheduleWithResultRef01Data["account_id"] = setup.idmap!!["account01"]
    accountUpdaterScheduleWithResultRef01Data["schedule_id"] = setup.idmap!!["schedule01"]

    val accountUpdaterScheduleWithResultRef01DataResult = accountUpdaterScheduleWithResultRef01Ent.create(accountUpdaterScheduleWithResultRef01Data, null)
    accountUpdaterScheduleWithResultRef01Data = Helpers.toMapAny(accountUpdaterScheduleWithResultRef01DataResult) ?: linkedMapOf()
    assertNotNull(accountUpdaterScheduleWithResultRef01Data, "expected create result to be a map")
    assertNotNull(accountUpdaterScheduleWithResultRef01Data["id"], "expected created entity to have an id")

    // LIST
    val accountUpdaterScheduleWithResultRef01Match = linkedMapOf<String, Any?>()
    accountUpdaterScheduleWithResultRef01Match["account_id"] = setup.idmap!!["account01"]
    accountUpdaterScheduleWithResultRef01Match["schedule_id"] = setup.idmap!!["schedule01"]

    val accountUpdaterScheduleWithResultRef01ListResult = accountUpdaterScheduleWithResultRef01Ent.list(accountUpdaterScheduleWithResultRef01Match, null)
    assertTrue(accountUpdaterScheduleWithResultRef01ListResult is List<*>,
        "expected list result to be an array, got " + accountUpdaterScheduleWithResultRef01ListResult)
    val accountUpdaterScheduleWithResultRef01List = accountUpdaterScheduleWithResultRef01ListResult as List<Any?>

    val foundItem = Struct.select(
        RunnerSupport.entityListToData(accountUpdaterScheduleWithResultRef01List),
        Struct.jm("id", accountUpdaterScheduleWithResultRef01Data["id"]))
    assertFalse(Struct.isempty(foundItem), "expected to find created entity in list")

  }

  @Test
  fun stream() {
    val streamingActive = linkedMapOf<String, Any?>(
      "feature" to linkedMapOf<String, Any?>(
        "streaming" to linkedMapOf<String, Any?>("active" to true),
      ),
    )
    val setup = accountUpdaterScheduleWithResultBasicSetup(streamingActive)
    Assumptions.assumeFalse(
      setup.live,
      "stream test streams the seeded fixture data (unit mode only)",
    )

    val ent = setup.client.accountUpdaterScheduleWithResult(null)
    val match = linkedMapOf<String, Any?>()

    // Materialised list result for the same op.
    val listedResult = ent.list(match, null)
    val listed = (listedResult as? List<Any?>) ?: emptyList<Any?>()

    // stream("list") yields items via the streaming feature's iterator.
    val streamed = ent.stream("list", match, null).toList()
    assertTrue(streamed.size > 0, "expected stream to yield items")
    assertEquals(listed.size, streamed.size, "expected stream to match list count")

    // Fallback: with streaming inactive, stream still yields the materialised
    // items.
    val setup2 = accountUpdaterScheduleWithResultBasicSetup(null)
    val ent2 = setup2.client.accountUpdaterScheduleWithResult(null)
    val streamed2 = ent2.stream("list", match, null).toList()
    assertEquals(listed.size, streamed2.size, "expected fallback stream to match list")
  }

  companion object {
    fun accountUpdaterScheduleWithResultBasicSetup(extra: MutableMap<String, Any?>?): RunnerSupport.EntityTestSetup {
      RunnerSupport.loadEnvLocal()

      val entityData: MutableMap<String, Any?>
      try {
        val entityDataSource = Files.readString(Paths.get(
            "..", ".sdk", "test", "entity", "account_updater_schedule_with_result", "AccountUpdaterScheduleWithResultTestData.json"))
        entityData = Helpers.toMapAny(Json.parse(entityDataSource)) ?: linkedMapOf()
      } catch (e: Exception) {
        throw AssertionError("failed to read account_updater_schedule_with_result test data: " + e.message, e)
      }

      val options = linkedMapOf<String, Any?>()
      options["entity"] = entityData["existing"]

      var client = BluefinPayconexSDK.testSDK(options, extra)

      // Generate idmap via transform, matching TS pattern.
      val idnames = mutableListOf<Any?>()
      idnames.add("account_updater_schedule_with_result01")
      idnames.add("account_updater_schedule_with_result02")
      idnames.add("account_updater_schedule_with_result03")
      idnames.add("account01")
      idnames.add("account02")
      idnames.add("account03")
      idnames.add("schedule01")
      idnames.add("schedule02")
      idnames.add("schedule03")
      val idmap = Struct.transform(idnames, Json.parse(
          "{\"`\$PACK`\": [\"\", {" +
          "\"`\$KEY`\": \"`\$COPY`\"," +
          "\"`\$VAL`\": [\"`\$FORMAT`\", \"upper\", \"`\$COPY`\"]" +
          "}]}"))

      // Detect ENTID env override before envOverride consumes it.
      val entidEnvRaw = RunnerSupport.getenv("BLUEFINPAYCONEX_TEST_ACCOUNT_UPDATER_SCHEDULE_WITH_RESULT_ENTID")
      val idmapOverridden = entidEnvRaw != null && entidEnvRaw.trim().startsWith("{")

      val envm = linkedMapOf<String, Any?>()
      envm["BLUEFINPAYCONEX_TEST_ACCOUNT_UPDATER_SCHEDULE_WITH_RESULT_ENTID"] = idmap
      envm["BLUEFINPAYCONEX_TEST_LIVE"] = "FALSE"
      envm["BLUEFINPAYCONEX_TEST_EXPLAIN"] = "FALSE"
      envm["BLUEFINPAYCONEX_APIKEY"] = "NONE"
      val env = RunnerSupport.envOverride(envm)

      var idmapResolved = Helpers.toMapAny(env["BLUEFINPAYCONEX_TEST_ACCOUNT_UPDATER_SCHEDULE_WITH_RESULT_ENTID"])
      if (idmapResolved == null) {
        idmapResolved = Helpers.toMapAny(idmap) ?: linkedMapOf()
      }

      val live = "TRUE" == env["BLUEFINPAYCONEX_TEST_LIVE"]
      if (live) {
        val liveOpts = linkedMapOf<String, Any?>()
        liveOpts["apikey"] = env["BLUEFINPAYCONEX_APIKEY"]
        val mergedOpts = Struct.merge(Struct.jt(liveOpts, extra))
        client = BluefinPayconexSDK(Helpers.toMapAny(mergedOpts))
      }

      val setup = RunnerSupport.EntityTestSetup()
      setup.client = client
      setup.data = entityData
      setup.idmap = idmapResolved
      setup.env = env
      setup.explain = "TRUE" == env["BLUEFINPAYCONEX_TEST_EXPLAIN"]
      setup.live = live
      setup.syntheticOnly = live && !idmapOverridden
      setup.now = System.currentTimeMillis()
      return setup
    }
  }
}

File: cpp/test/account_updater_schedule_with_result_entity_test.cpp
// Generated basic-flow test for the account_updater_schedule_with_result entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct AccountUpdaterScheduleWithResultSetup {
  std::shared_ptr<BluefinPayconexSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static AccountUpdaterScheduleWithResultSetup account_updater_schedule_with_result_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/account_updater_schedule_with_result/AccountUpdaterScheduleWithResultTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinPayconexSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("account_updater_schedule_with_result01"), Value("account_updater_schedule_with_result02"), Value("account_updater_schedule_with_result03"), Value("account01"), Value("account02"), Value("account03"), Value("schedule01"), Value("schedule02"), Value("schedule03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINPAYCONEX_TEST_ACCOUNT_UPDATER_SCHEDULE_WITH_RESULT_ENTID", idmap},
    {"BLUEFINPAYCONEX_TEST_LIVE", Value("FALSE")},
    {"BLUEFINPAYCONEX_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINPAYCONEX_TEST_ACCOUNT_UPDATER_SCHEDULE_WITH_RESULT_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINPAYCONEX_TEST_LIVE") == Value("TRUE");

  AccountUpdaterScheduleWithResultSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void account_updater_schedule_with_result_entity_instance() {
  auto testsdk = BluefinPayconexSDK::testSDK();
  auto ent = testsdk->account_updater_schedule_with_result();
  ASSERT_EQ(ent->getName(), std::string("account_updater_schedule_with_result"), "entity name");
}

static void account_updater_schedule_with_result_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"account_updater_schedule_with_result", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinPayconexSDK::testSDK(seed, sdkopts);
  auto se = strsdk->account_updater_schedule_with_result();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinPayconexSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->account_updater_schedule_with_result();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void account_updater_schedule_with_result_entity_basic() {
  auto setup = account_updater_schedule_with_result_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create", "list"}) {
    auto sk = is_control_skipped("entityOp", std::string("account_updater_schedule_with_result.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto account_updater_schedule_with_result_ref01_ent = client->account_updater_schedule_with_result();
  Value account_updater_schedule_with_result_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "account_updater_schedule_with_result"}), "account_updater_schedule_with_result_ref01"));
  if (!account_updater_schedule_with_result_ref01_data.is_map()) account_updater_schedule_with_result_ref01_data = vmap();
  setp(account_updater_schedule_with_result_ref01_data, "account_id", getp(setup.idmap, "account01"));
  setp(account_updater_schedule_with_result_ref01_data, "schedule_id", getp(setup.idmap, "schedule01"));
  {
    Value account_updater_schedule_with_result_ref01_data_result = account_updater_schedule_with_result_ref01_ent->create(Struct::clone(account_updater_schedule_with_result_ref01_data), Value::undef());
    account_updater_schedule_with_result_ref01_data = Helpers::toMapAny(account_updater_schedule_with_result_ref01_data_result);
    if (!account_updater_schedule_with_result_ref01_data.is_map()) account_updater_schedule_with_result_ref01_data = vmap();
    ASSERT_TRUE(account_updater_schedule_with_result_ref01_data.is_map(), "expected create result to be a map");
    ASSERT_TRUE(!getp(account_updater_schedule_with_result_ref01_data, "id").is_undef(), "expected created entity to have an id");
  }

  // LIST
  Value account_updater_schedule_with_result_ref01_match = vmap();
  setp(account_updater_schedule_with_result_ref01_match, "account_id", getp(setup.idmap, "account01"));
  setp(account_updater_schedule_with_result_ref01_match, "schedule_id", getp(setup.idmap, "schedule01"));
  Value account_updater_schedule_with_result_ref01_list = account_updater_schedule_with_result_ref01_ent->list(Struct::clone(account_updater_schedule_with_result_ref01_match), Value::undef());
  ASSERT_TRUE(account_updater_schedule_with_result_ref01_list.is_list(), "expected list result to be an array");
  {
    std::vector<Value> found = Struct::select(entity_list_to_data(account_updater_schedule_with_result_ref01_list), vmap({{"id", getp(account_updater_schedule_with_result_ref01_data, "id")}}));
    ASSERT_TRUE(!found.empty(), "expected to find created entity in list");
  }

}

int main() {
  T_RUN(account_updater_schedule_with_result_entity_instance);
  T_RUN(account_updater_schedule_with_result_entity_stream);
  T_RUN(account_updater_schedule_with_result_entity_basic);
  return sdktest::summary("account_updater_schedule_with_result_entity_test");
}
